src/hugepages.c

```diff
@@ -0,0 +1,286 @@
+static const char NODE[] = "node";
+static const char HUGEPAGES_DIR[] = "hugepages";
+static const char SYS_NODE_HUGEPAGES[] = "/sys/devices/system/node/%s/hugepages";
+static const char SYS_MM_HUGEPAGES[] = "/sys/kernel/mm/hugepages";
+static const char CONFIG_NAME[] = "hugepages";
+static const char CFG_ENA_NUMA[] = "EnableNuma";
+static const char CFG_ENA_MM[] = "EnableMM";
+
+static const char *CONFIG_KEYS[] = {
+  CFG_ENA_NUMA,
+  CFG_ENA_MM,
+};
+static const size_t CONFIG_KEYS_NUM = sizeof(CONFIG_KEYS)/sizeof(*CONFIG_NAME);
+static int g_config_ena_numa = 1;
+static int g_config_ena_mm = 1;
+
+static int huge_config_callback (const char *key, const char *val)
+{
+  INFO("HugePages config key='%s', val='%s'", key, val);
+
+  if (0 == strcasecmp(key, CFG_ENA_NUMA)) {
+    g_config_ena_numa = IS_TRUE(val);
+    return 0;
+  }
+  if (0 == strcasecmp(key, CFG_ENA_MM)) {
+    g_config_ena_mm = IS_TRUE(val);
+    return 0;
+  }
+
+  return -1;
+}
+
+static void submit_one (const char *plug_inst, const char *type,
+  const char *type_instance, derive_t value)
+{
+  value_t values[1];
+  value_list_t vl = VALUE_LIST_INIT;
+
+  values[0].derive = value;
+
+  vl.values = values;
+  vl.values_len = 1;
+  sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+  sstrncpy (vl.plugin, PLUGIN_NAME, sizeof (vl.plugin));
+  sstrncpy (vl.plugin_instance, plug_inst, sizeof (vl.plugin_instance));
+  sstrncpy (vl.type, type, sizeof (vl.type));
+
+  if (type_instance != NULL) {
+    sstrncpy (vl.type_instance, type_instance, sizeof (vl.type_instance));
+  }
+
+  DEBUG("submit_one pl_inst:%s, inst_type %s, type %s, val=%"PRIu64"",
+      plug_inst, type_instance, type, value);
+
+  plugin_dispatch_values (&vl);
+}
+
+static int read_hugepage_entry(const char *path, const char* entry,
+    const char* plinst, const char* hpsize)
+{
+  char path2[512];
+  long value = 0;
+  snprintf(path2, sizeof(path2), "%s/%s", path, entry);
+
+  FILE *fh = fopen(path2, "rt");
+  if (NULL == fh) {
+    ERROR("Cannot open %s", path2);
+    return -1;
+  }
+
+  if (fscanf(fh, "%ld", &value) !=1) {
+    ERROR("Cannot parse file %s", path2);
+    fclose(fh);
+    return -1;
+  }
+
+  submit_one (plinst, entry, hpsize, value);
+
+  fclose(fh);
+  return 0;
+}
+
+static int read_syshugepage_dir(const char* path, const char* dirhpsize,
+    const char* node)
+{
+  DIR       *dir = NULL;
+  struct dirent *entry = NULL;
+  struct dirent *result = NULL;
+  size_t name_max = 0;
+  size_t len = 0;
+
+  dir = opendir(path);
+  if (NULL == dir) {
+    ERROR("Cannot open directory %s", path);
+    return -1;
+  }
+
+  name_max = pathconf(path, _PC_NAME_MAX);
+  if (name_max == -1) {    /* Limit not defined, or error */
+    name_max = 255;     /* Take a guess */
+  }
+
+  len = offsetof(struct dirent, d_name) + name_max + 1;
+  entry = malloc(len);
+  if (entry == NULL) {
+    ERROR("Malloc returned NULL");
+    return -1;
+  }
+
+  while (0 == readdir_r(dir, entry, &result)) {
```

Comment:
readdir_r has become deprecated in glibc 2.24. Since we report warnings as errors by default, this will fail on newer platforms. See http://man7.org/linux/man-pages/man3/readdir_r.3.html for the details.